crates/storage/nippy-jar/src/writer.rs

```diff
@@ -0,0 +1,341 @@
+                let num_offsets = (length - 1) / bytes_per_offset;
+                let new_num_offsets = num_offsets.saturating_sub(remaining_to_prune as u64);
+
+                // If all rows are to be pruned
+                if new_num_offsets <= 1 {
+                    // <= 1 because the one offset would actually be the expected file data size
+                    self.offsets_file.set_len(1)?;
+                    self.data_file.set_len(0)?;
+                } else {
+                    // Calculate the new length for the on-disk offset list
+                    let new_len = 1 + new_num_offsets * bytes_per_offset;
+                    // Seek to the position of the last offset
+                    self.offsets_file
+                        .seek(SeekFrom::Start(new_len.saturating_sub(bytes_per_offset)))?;
+                    // Read the last offset value
+                    let mut last_offset = [0u8; 8];
+                    self.offsets_file.read_exact(&mut last_offset)?;
+                    let last_offset = u64::from_le_bytes(last_offset);
+
+                    // Update the lengths of both the offsets and data files
+                    self.offsets_file.set_len(new_len)?;
+                    self.data_file.set_len(last_offset)?;
+                }
+            } else {
+                // If the offsets file only contains the byte representation
+                self.offsets_file.set_len(1)?;
+                self.data_file.set_len(0)?;
+            }
+        }
+
+        self.offsets_file.seek(SeekFrom::End(0))?;
+        self.data_file.seek(SeekFrom::End(0))?;
+
+        self.jar.rows = self.jar.rows.saturating_sub(num_rows);
+        if self.jar.rows == 0 {
+            self.jar.max_row_size = 0;
+        }
+        self.jar.freeze_config()?;
+
+        Ok(())
+    }
+
+    /// Updates [`NippyJar`] with the new row count and maximum uncompressed row size, while
+    /// resetting internal fields.
+    fn finalize_row(&mut self) {
+        self.jar.max_row_size = self.jar.max_row_size.max(self.uncompressed_row_size);
+        self.jar.rows += 1;
+
+        self.tmp_buf.clear();
+        self.uncompressed_row_size = 0;
+        self.column = 0;
+    }
+
+    /// Commits configuration and offsets to disk. It drains the internal offset list.
+    pub fn commit(&mut self) -> Result<(), NippyJarError> {
+        self.commit_offsets()?;
+
+        // Flushes `max_row_size` and total `rows` to disk.
+        self.jar.freeze_config()?;
+
+        Ok(())
+    }
+
+    /// Flushes offsets to disk.
+    pub(crate) fn commit_offsets(&mut self) -> Result<(), NippyJarError> {
+        // The last offset on disk can be the first offset on `self.offset_list` given how
+        // `append_column()` works alongside commit. So we need to skip it.
+        let mut last_offset_ondisk = None;
+
+        if self.offsets_file.metadata()?.len() > 1 {
+            self.offsets_file.seek(SeekFrom::End(-8))?;
```

Comment:
can we move this `8` to const as a size of offset? I see it throughout the crate, may be confusing